### PR TITLE
feat(prisma): add prisma adapter

### LINKED ISSUE #80: Add prisma adapter
Add prisma.io adapter, if possible.

### PR DESCRIPTION
**Pull request by daniharo**
Prisma adapter added taking [TypeORM adapter](https://github.com/grammyjs/storages/tree/main/packages/typeorm) as example.

TODO:

- [x] run tests (can't run because `@grammyjs/storage-utils` doesn't seem to be imported correctly)
- [x] create PR for grammyjs/website: https://github.com/grammyjs/website/pull/440

Closes #80

### CODE REVIEW

File: packages/prisma/LICENSE
Comment (KnorpelSenf) [inline packages/prisma/LICENSE:3]: It's probably good to licence your own code under your own name

File: packages/prisma/src/index.ts
Comment (KnorpelSenf) [inline packages/prisma/src/index.ts:1]:
```suggestion
import { type StorageAdapter } from 'grammy';
```

**Review by KnorpelSenf — CHANGES_REQUESTED** [review]
Thank you so much!

I'll leave a proper review to @Satont, here's just a short comment about the licensing, as well as a nitpick that lets you drop grammY as a runtime dependency.

**Review by KnorpelSenf — APPROVED** [review]
LGTM, maintainer will merge after reviewing

File: packages/prisma/src/index.ts
Comment (Satont) [reply packages/prisma/src/index.ts]: Whats difference with `import type { StorageAdapter  }`?

**Comment by Satont** [comment]
@daniharo can you try build utils manually and then run the tests?

**Comment by Satont** [comment]
For the tests, you have to call `db push` for prisma, before each of the cases. Can you do this via `exec` so you don't have to do it by hand, I assume?

Also, I think you have to create a new base before each of the cases (delete the old one and create a new one and do the push). You can use `beforeEach` hook from `vitest` package.

**Comment by Satont** [comment]
Please add adapter to the `packages/README.md` list.

**Comment by Satont** [comment]
Rebase with master please. I added running tests on PR.

**Comment by Satont** [comment]
Ok i see tests failed due indepented to pr reasons, so we can ignore that. Just run them manually until i'll fix tests.

**Comment by daniharo** [comment]
> If you not familiar with tests enough i can fix them myself, if you wan't'

I got the tests working, just have to fix one that fails.

**Comment by daniharo** [comment]
@Satont I'm getting this error inside the `beforeEach` calls
```
Error: 
Invalid `prisma.session.deleteMany()` invocation:


  Error occurred during query execution:
ConnectorError(ConnectorError { user_facing_error: None, kind: ConnectionError(Timed out during query execution.) })
```

Please check if you can solve it

**Comment by Satont** [comment]
@daniharo that seems like sqlite database locked. You can use postgres with `pg-mem` like in [psql](https://github.com/grammyjs/storages/blob/main/packages/psql/__tests__/node.ts) storage test instead.

**Comment by daniharo** [comment]
> @daniharo that seems like sqlite database locked. You can use postgres with `pg-mem` like in [psql](https://github.com/grammyjs/storages/blob/main/packages/psql/__tests__/node.ts) storage test instead.

Prisma needs a connection URL to create the client, I don't think `pg-mem` gives that possibility

**Comment by Satont** [comment]
> > @daniharo that seems like sqlite database locked. You can use postgres with `pg-mem` like in [psql](https://github.com/grammyjs/storages/blob/main/packages/psql/__tests__/node.ts) storage test instead.
> 
> Prisma needs a connection URL to create the client, I don't think `pg-mem` gives that possibility.
> 
> Gonna try with [mongodb-memory-server](https://github.com/nodkz/mongodb-memory-server)

It gives, but you need to run `push` into code, not via cli.

**Comment by daniharo** [comment]
Tests are now passing.

I just had to add `threads: false` to vitest config in this package to make it work with SQLite (commit https://github.com/grammyjs/storages/pull/108/commits/0dc5550ab41dff242dec30521a1e420c46d5e1a6)

**Comment by Satont** [comment]
So, it is LGTM except version of package. Can you set it to 2.0.0 to respect other packages?

**Review by Satont — APPROVED** [review]

**Comment by daniharo** [comment]
I think this is now ready for merging

**Comment by KnorpelSenf** [comment]
Would you mind updating the docs on the website and linking the pull request here?

**Comment by daniharo** [comment]
> Would you mind updating the docs on the website and linking the pull request here?

Done

**Comment by KnorpelSenf** [comment]
Thanks! The website is already updated.

I don't know what the errors in CI are, but I assume that @Satont will know.

**Pull request merged**
